hawkular-datamining-api/src/main/java/org/hawkular/datamining/api/base/InMemorySubscriptionManager.java

```diff
@@ -74,8 +67,9 @@ public void setPredictionListener(PredictionListener predictionListener) {
     }
 
     @Override
-    public void subscribe(Subscription subscription) {
+    public synchronized void subscribe(Subscription subscription) {
```

Comment:
Yes, that is the  purpose of synchronizing it, subscription can get lost. 

I'm not sure about `unsubscribe`, I think not synchronized version can't cause race condition. However it cause this 
1. `subscribe(sub1)`
2. `unsubscribe(sub1)` -> exception not found (getting historical metrics can take some time)
What do you think about this? Then `isSubscribed` and `subscription` should be synchronized too.